### ISSUE TITLE
Irrelevant double spend sometimes leaves wallet in an inconsistent state

### ISSUE THREAD

**Issue by saulbein** [bitcoinj/bitcoinj#1374]
# Situation

Wallet receives 2 relevant transactions from the network: tx2 & tx3 (spending tx0 & tx1).
Then tx4 is received, which double spends tx3 and also spends an irrelevant output of tx2.
Since tx4 is irrelevant it does not get saved to the wallet, however, tx2's output is still marked as spent by tx4. When reloading the wallet this causes an inconsistent state as tx2's output cannot point spentBy back to tx4.

# Solutions

2 likely solutions could be implemented for this problem: 
1. Saving the irrelevant double spend (tx4)
2. Disconnecting the irrelevant output (tx2 -> tx4).

# Replication

I've added this test to WalletTest in order to have a (somewhat) minimal way of replicating the issue:

```
    @Test
    public void testIrrelevantDoubleSpend() throws Exception {
        Transaction tx0 = createFakeTx(PARAMS);
        Transaction tx1 = createFakeTx(PARAMS);

        Transaction tx2 = new Transaction(PARAMS);
        tx2.addInput(tx0.getOutput(0));
        tx2.addOutput(COIN, myAddress);
        tx2.addOutput(COIN, OTHER_ADDRESS);

        sendMoneyToWallet(AbstractBlockChain.NewBlockType.BEST_CHAIN, tx2, tx1, tx0);

        // tx3 and tx4 double spend each other
        Transaction tx3 = new Transaction(PARAMS);
        tx3.addInput(tx1.getOutput(0));
        tx3.addOutput(COIN, myAddress);
        tx3.addOutput(COIN, OTHER_ADDRESS);
        wallet.receivePending(tx3, null);

        // tx4 also spends irrelevant output from tx2
        Transaction tx4 = new Transaction(PARAMS);
        tx4.addInput(tx1.getOutput(0)); // spends same output
        tx4.addInput(tx2.getOutput(1));
        tx4.addOutput(COIN, OTHER_ADDRESS);

        // tx4 does not actually get added to wallet here since it by itself is irrelevant
        sendMoneyToWallet(AbstractBlockChain.NewBlockType.BEST_CHAIN, tx4);

        // since tx4 is not saved, tx2 output 1 will have bad spentBy
        wallet = roundTrip(wallet);

        assertTrue(wallet.isConsistent());
    }
```
I also have a workaround/solution implementation by disconnecting irrelevant outputs, will put a PR soon.

**Comment by saulbein** [comment]
Added PR #1375.

**Comment by schildbach** [comment]
Very interesting find! Many thanks for debugging this.

**Comment by schildbach** [comment]
Generally this issue seems to be really rare. I haven't seen any wallet inconsistency problems since the last wallet serialization issue was fixed. May I ask did you stumble upon it by coincidence or is there a specific usecase that fosters this issue?

**Comment by jdojff** [comment]
Hi, this was noticed in the livenet, and it does look like a reasonable usecase.

I think we need one more update for similar usecase:
Received irrelevant double spend for dead transaction.
`isTransactionRelevant` - says it is relevant
In `processTxFromBestChain` check of `findDoubleSpendsAgainst(tx, pending);` - says irrelevant

Fix: check against pending pool in `isTransactionRelevant` method.
Not sure, though, about reorganizing and possible other issues?

```
    /**
     * <p>Returns true if the given transaction sends coins to any of our keys, or has inputs spending any of our outputs,
     * and also returns true if tx has inputs that are spending outputs which are
     * not ours but which are spent by pending transactions.</p> // <-------- this tells that we will check pending transactions
     *
     * <p>Note that if the tx has inputs containing one of our keys, but the connected transaction is not in the wallet,
     * it will not be considered relevant.</p>
     */
    public boolean isTransactionRelevant(Transaction tx) throws ScriptException {
        lock.lock();
        try {
            return tx.getValueSentFromMe(this).signum() > 0 ||
                   tx.getValueSentToMe(this).signum() > 0 ||
                   !findDoubleSpendsAgainst(tx, transactions).isEmpty();   // <-------- this checks all transactions. Should check against pending??
        } finally {
            lock.unlock();
        }
}
```
and
```
 // Kill txns in conflict with this tx
Set<Transaction> doubleSpendTxns = findDoubleSpendsAgainst(tx, pending); // <-------- this checks pending transactions
if (!doubleSpendTxns.isEmpty()) {
```

**Comment by saulbein** [comment]
I've included another way to fix the issue referenced by @jdojff - it adds just adds another check to see if the tx is a double spend against any transactions in the wallet and disconnect irrelevant inputs if that is the case (side effect of this was to change that `findDoubleSpendsAgainst` leaves out the transaction in question, which made some payent channel tests fail).

However, I think @jdojff solution might be better if we can confirm that all the edge cases are covered (for example what if the transaction in question is in a block that starts a reorg?).

**Comment by schildbach** [comment]
Fixed by #1375.
